File: 2015/12/javascript/main.mjs
import fs from "node:fs";

// --- Constants ---
const INPUT_FILE = "../input/input.txt";
const REGEX = /(-?\d+)/g;

// --- Functions ---
function sum(json) {
	const type = typeof json;
	if (type === "number") return Number(json);
	if (type === "object") {
		if (Array.isArray(json)) return json.reduce((acc, x) => acc + sum(x), 0);
		if (Object.values(json).includes("red")) return 0;
		return Object.values(json).reduce((acc, x) => acc + sum(x), 0);
	}
	return 0;
}

// --- Parse the input file ---
const text = fs.readFileSync(INPUT_FILE, 'utf8').trim();

// --- Puzzle 1 ---
let total = text.match(REGEX).reduce((acc, x) => acc + Number(x), 0);
console.log("1. Sum of all numbers: ", total.toLocaleString());

// --- Puzzle 2 ---
const json = JSON.parse(text);
total = sum(json);
console.log("2. Sum of all numbers: ", total.toLocaleString());

File: 2015/12/python/main.py
import json
import re

# --- Constants --
INPUT_FILE = "../input/input.txt"
REGEX = r"(-?\d+)"

# --- Functions ---
def sum(js):
    t = type(js)
    if t == int:
        return js
    elif t == list:
        total = 0
        for v in js: total += sum(v)
        return total
    elif t == dict:
        values = list(js.values())
        try:
            values.index("red")
            return 0
        except ValueError:
            total = 0
            for v in values: total += sum(v)
            return total
    return 0

# --- Read and Parse the input file --
file = open(INPUT_FILE, "r")
text = file.read().strip()
file.close()

# --- Puzzle 1 ---
total = 0
for cap in re.findall(REGEX, text): total += int(cap)
print(f"1. Sum of all numbers: {total:,d}")

# --- Puzzle 2 ---
total = sum(json.loads(text))
print(f"2. Sum of all numbers: {total:,d}")
